feat: Add image caption functionality with encryption and PII scanning

dagbolade opened the pull request:
##  **Fixes Issue #2**

**Link**: https://github.com/IEEE-VIT/priv-guard/issues/2#issue-3280108679

##  **Problem Statement**
Users needed the ability to add custom captions or short descriptions to images to help organize and search their media later, similar to how posts are made on social media platforms.

##  **Solution Implemented**

### 🔹 **Caption Input System**
- **Caption Dialog**: Added text input dialog when uploading or editing images
- **User Experience**: Clean, intuitive interface that prompts users to add context to their images
- **Edit Capability**: Users can modify captions of existing images through long-press menu

### 🔹 **Secure Storage Architecture**
- **AES Encryption**: All caption metadata is encrypted using AES-CBC encryption with secure key management
- **Secure Storage**: Uses Flutter Secure Storage for encryption keys and encrypted file storage
- **Unique Naming**: Timestamp-based storage names prevent file collisions and ensure data integrity

### 🔹 **Search & Organization Features**
- **Real-time Search**: Added search bar to filter images by caption content instantly
- **Alphabetical Sorting**: Toggle between A-Z and Z-A sorting based on caption text
- **Dynamic Filtering**: Search results update as users type, providing immediate feedback

### 🔹 **Privacy Integration**
- **PII Detection**: Caption text is automatically processed through the existing TFLite model
- **Privacy Scanning**: Captions are scanned for sensitive information (phone numbers, emails, etc.)
- **Seamless Integration**: Works with existing privacy analysis pipeline without breaking current functionality

##  **Technical Implementation**

### **Files Modified:**
1. **lib/screens/gallery_screen.dart**
   - Added caption input dialog (_promptForCaption())
   - Implemented encrypted metadata storage (_saveImageMetadata(), _loadImageMetadata())
   - Added search functionality with real-time filtering
   - Added sort capability by caption text
   - Enhanced image deletion to clean up metadata

2. **lib/screens/scan_analysis_screen.dart**
   - Integrated caption text with PII detection pipeline
   - Added caption processing in image analysis workflow

3. **test/widget_test.dart**
   - Fixed test compatibility issues

### **Key Features:**
-  Caption input when uploading images
-  AES-encrypted metadata storage
-  Real-time search by caption
-  Alphabetical sorting functionality
-  PII scanning integration
-  Secure file management
-  Edit existing captions
-  Clean metadata deletion

##  **Security & Privacy**
- All caption data is encrypted before storage
- Uses industry-standard AES-CBC encryption
- Secure key management through Flutter Secure Storage
- Caption text processed through existing PII detection system
- No sensitive data stored in plain text

##  **Testing**
- Code passes Flutter static analysis
- No compilation errors
- All core functionality verified
- Maintains compatibility with existing features

##  **User Experience**
Users can now:
1. Add captions when uploading images
2. Search their image collection by caption text
3. Sort images alphabetically by caption
4. Edit captions of existing images
5. Have their privacy protected through automatic PII scanning

---

##  **Maintainer Note**
Please add the following labels to this PR:
- hacktoberfest-accepted
- hacktoberfest2025-accepted

This contribution solves the caption functionality requested in issue #2 and maintains the app's privacy-first approach.

dagbolade committed:
feat: Add image caption functionality with encryption and PII scanning

- Add caption input dialog when uploading/editing images
- Implement AES-encrypted metadata storage with secure key management
- Add real-time search functionality to filter images by caption
- Add sort functionality to organize images alphabetically by caption
- Integrate caption text with existing TFLite PII detection pipeline
- Use unique timestamp-based storage names to prevent file collisions
- Add secure metadata deletion when images are removed

Fixes #2

Shaurya-Saini commented:
Hi @dagbolade, thanks for contributing to the project! The changes look good overall, but I found a few issues that need to be addressed before merging:

---

## 1. RenderFlex Overflow Error

* **Error:**

  ```
  A RenderFlex overflowed by 7.0 pixels on the bottom.
  ```
  **Widget:** `Column`  
  **File:** `lib/screens/main_navigation_screen.dart:21`

* **Details:**
  This error occurs after saving an edited or newly created caption for posts.

* Fix the layout to prevent overflow (e.g., wrap with `SingleChildScrollView` or use `Expanded` widgets)
* Test on devices with smaller screen heights to confirm fix

---

## 2. Caption Feature Behavior

* **Issue:** The option to add/edit captions is currently also available for **text posts**, which should not be the case.
  Captions should only be supported for **image posts**.

* Restrict caption-related options (add/edit) to image posts only

---

## 3. Search Bar Reset Issue

* **Issue:**
  After performing a search, there’s **no option to cancel or clear** the search filter.
  Currently, the app must be restarted to reset results.

  > This issue will be explicitly addressed in the upcoming update for **ISSUE #1**, so it can be skipped in this PR.

---

### Summary

Please fix **issues 1 and 2** mentioned above in this PR review.
The **search bar (issue 3)** can be handled later as part of the ISSUE #1.

Once the changes are pushed, I’ll re-review and approve the merge.

Thanks again for your contribution! 🙌

dagbolade committed:
fix: Address PR review feedback

- Fix RenderFlex overflow by wrapping body with SafeArea
- Restrict caption feature to image posts only
- Remove caption option from text posts in context menu

Fixes issues #1 and #2 from PR review comment

dagbolade commented:
Hi @Shaurya-Saini  can you review it thanks

Shaurya-Saini commented:
Hi @dagbolade, thanks again for your quick updates. The earlier issues are mostly resolved, but I noticed a new runtime error related to the caption dialog that needs attention before merging.

---

### Issue — `TextEditingController` Used After Disposal

File: `lib/screens/gallery_screen.dart:997`
Location: Inside the `_promptForCaption` method
Error message:

```
A TextEditingController was used after being disposed.
Once you have called dispose() on a TextEditingController, it can no longer be used.
```

This issue is happening because the `TextEditingController` in `_promptForCaption` is being disposed too early, immediately after the dialog is closed. When the dialog closes, Flutter might still be running some pending animations or UI cleanup tasks related to the `TextField`. If the controller is disposed before those updates finish, Flutter tries to access a controller that no longer exists, causing this error.

---

### Recommended Fix
The disposal of the controller should be slightly delayed or managed more safely to ensure it happens after all dialog animations and UI updates finish. This can be done by deferring the disposal until the current frame has completed or by moving the controller’s lifecycle management into the widget’s State, where it’s created once and disposed properly in the widget’s dispose() method. Either approach ensures the controller isn’t used after being disposed and prevents this runtime error.

---

Please look into this and update the PR once fixed 
Everything else seems good to go after this.

dagbolade committed:
Defer disposal until after frame completes

dagbolade commented:
Hi! I've fixed the TextEditingController disposal issue in the `_promptForCaption` method.

**What was the problem:**
The controller was being disposed immediately after the dialog closed, but Flutter was still running animations and UI cleanup tasks that needed to access it.

**The fix:**
I wrapped the disposal logic in `WidgetsBinding.instance.addPostFrameCallback((_) { ... })` to defer the disposal until after the current frame completes. This ensures all dialog animations and UI updates finish before the controller is disposed.

This is the same pattern already used successfully in the `_showCreateTextPostDialog` method, so it's consistent with the existing codebase.

The runtime error should now be resolved. Ready for review! @Shaurya-Saini
